addons/mail/static/src/service_worker.js

```diff
@@ -13,6 +15,28 @@ const PUSH_NOTIFICATION_ACTION = {
     DECLINE: "DECLINE",
 };
 
+const rtcLogs = {
+    _maxLength: 5,
+    _timelines: [],
+    _snapshots: [],
+    add(type, entry, value) {
+        let arr = this._snapshots;
+        if (type === "timeline") {
+            arr = this._timelines;
+        }
+        arr.push({ [entry]: value });
+        if (arr.length > this._maxLength) {
+            arr.shift();
+        }
+    },
+    getLogs() {
+        return {
+            timelines: Object.assign({}, ...this._timelines),
+            snapshots: Object.assign({}, ...this._snapshots),
+        };
+    },
+};
```

Comment:
This should be saved in shared worker, not service worker.

While they have many similarities like 1 per origin, there is a massive difference to consider here:

- shared worker stays alive as long as there is at least 1 tab. It is killed immediately when all tabs are closed.
- service worker can be killed somewhat arbitrarily and revive by the browser, even when all tabs are removed.

Storing data in service worker is not a good idea, because this can be arbitrarily lost. Shared worker on the other hand is guaranteed to stay alive when the web app stays in used.

Reply:
@alexkuhn ok I will make the PR for master, I think it will need a diff a bit too rich (like extra assets) to be merged in 18.2